protocol/message_persistence.go

```diff
@@ -2382,16 +2383,13 @@ func getMessagesFromScanRows(db sqlitePersistence, rows *sql.Rows, withCursor bo
 
 		if msg, ok := messageIdx[message.ID]; !ok {
 			messageIdx[message.ID] = &message
+			messages = append(messages, &message)
 		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
 			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
 			messageIdx[message.ID] = msg
```

Comment:
The current implementation doesn't update a `message` in `messages`. It overwrites the pointer in `messageIdx[message.ID]`. But what we need to do is update the value of the pointer and not the pointer itself.

```suggestion
			*messageIdx[message.ID] = *msg
```

Reply:
I think current implementation will update a message in messages? also i think we don't even need the line?
```
messageIdx[message.ID] = msg
```

Reply:
We need that line because otherwise we'll never update the message's payload in case there's multiple rows for that message. That's what I tried pointing out here: https://github.com/status-im/status-go/pull/2991#discussion_r1036038644


Let's play this through and imagine there's 2 rows for this message, so we'll pass through this snippet here twice:
```golang
		if msg, ok := messageIdx[message.ID]; !ok {
			messageIdx[message.ID] = &message
			messages = append(messages, &message)
		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
			messageIdx[message.ID] = msg
	        }
```

1. In the first round, `messageIdx[message.iD]` doesn't exist, so we end up in that first `if` block and store the message in the map (we store it there, so we can quickly access it in the second row, to update its payload). We then add our `message` to the list of messages to be returned.
2. In the second round, the message now does exists in `messageIdx` so we endup in the `else if` block. We extract the message paylaod from the second row and update the `msg` in our `messageIdx` such that it includes that payload.(This is the part we can not leave out, otherwise we'll return incomplete message data. It's crucial that the `msg` that lives in `messages` gets that same update.

As per @Samyoul 's comment, if we do

```golang
messageIdx[message.ID] = msg
```

In we'll overwrite the pointer in `messagesIdx` for `msg`, which is the same that lives in `messages`!
We want instead update the **value** of the pointer hence:

```golang
messageIdx[message.ID] = *msg
```

This will ensure the message in `messages` will receive that update of the second round.

Hope this makes sense..

Reply:
> emm.. I think current implementation will update a message in messages? also i think we don't even need the line?

@qfrank, I apologise I see that you are correct. `msg` is already a pointer to the active `common.Messenger`

**Sample Code**

```go
type discordMsg struct {
	Data []byte
}
type msge struct {
	ID         string
	Payload    []byte
	DiscordMsg *discordMsg
}

func (m *msge) GetDiscordMessage() *discordMsg {
	return m.DiscordMsg
}

func getUpdatedChatMessagePayload(original, attachment *discordMsg) []byte {
	original.Data = append(original.Data, attachment.Data...)
	return original.Data
}

func Test(t *testing.T) {
	var messages []*msge
	messageIdx := make(map[string]*msge, 0)

	data := []msge{
		{ID: "123"},
		{ID: "234"},
		{ID: "345"},
		{ID: "456", DiscordMsg: &discordMsg{Data: []byte{0xff, 0xfe, 0xfd}}},
		{ID: "456", DiscordMsg: &discordMsg{Data: []byte{0xfc, 0xfb, 0xfa}}},
		{ID: "456", DiscordMsg: &discordMsg{Data: []byte{0xf9, 0xf8, 0xf7}}},
	}

	for i := range data {
		if msg, ok := messageIdx[data[i].ID]; !ok {
			messageIdx[data[i].ID] = &data[i]
			messages = append(messages, &data[i])
		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
			msg.Payload = getUpdatedChatMessagePayload(discordMessage, data[i].GetDiscordMessage())
		}
	}

	spew.Dump(messageIdx, messages)
}
```

Output
```
(map[string]*server.msge) (len=4) {
 (string) (len=3) "345": (*server.msge)(0xc0005202a0)({
  ID: (string) (len=3) "345",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 }),
 (string) (len=3) "456": (*server.msge)(0xc0005202d0)({
  ID: (string) (len=3) "456",
  Payload: ([]uint8) (len=9 cap=16) {
   00000000  ff fe fd fc fb fa f9 f8  f7                       |.........|
  },
  DiscordMsg: (*server.discordMsg)(0xc00050ef18)({
   Data: ([]uint8) (len=9 cap=16) {
    00000000  ff fe fd fc fb fa f9 f8  f7                       |.........|
   }
  })
 }),
 (string) (len=3) "123": (*server.msge)(0xc000520240)({
  ID: (string) (len=3) "123",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 }),
 (string) (len=3) "234": (*server.msge)(0xc000520270)({
  ID: (string) (len=3) "234",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 })
}
([]*server.msge) (len=4 cap=4) {
 (*server.msge)(0xc000520240)({
  ID: (string) (len=3) "123",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 }),
 (*server.msge)(0xc000520270)({
  ID: (string) (len=3) "234",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 }),
 (*server.msge)(0xc0005202a0)({
  ID: (string) (len=3) "345",
  Payload: ([]uint8) <nil>,
  DiscordMsg: (*server.discordMsg)(<nil>)
 }),
 (*server.msge)(0xc0005202d0)({
  ID: (string) (len=3) "456",
  Payload: ([]uint8) (len=9 cap=16) {
   00000000  ff fe fd fc fb fa f9 f8  f7                       |.........|
  },
  DiscordMsg: (*server.discordMsg)(0xc00050ef18)({
   Data: ([]uint8) (len=9 cap=16) {
    00000000  ff fe fd fc fb fa f9 f8  f7                       |.........|
   }
  })
 })
}
```

Reply:
Sorry for the late update here. Yes @qfrank and I have discussed this offline as well.
Thanks @Samyoul for the detailed reproducible!